synapse/storage/databases/main/user_directory.py

```diff
@@ -811,94 +811,132 @@ async def search_user_dir(
+                        AND vector @@ to_tsquery('simple', ?)
+                    ORDER BY
+                        (CASE WHEN d.user_id IS NOT NULL THEN 4.0 ELSE 1.0 END)
+                        * (CASE WHEN display_name IS NOT NULL THEN 1.2 ELSE 1.0 END)
+                        * (CASE WHEN avatar_url IS NOT NULL THEN 1.2 ELSE 1.0 END)
+                        * (
+                            3 * ts_rank_cd(
+                                '{0.1, 0.1, 0.9, 1.0}',
+                                vector,
+                                to_tsquery('simple', ?),
+                                8
+                            )
+                            + ts_rank_cd(
+                                '{0.1, 0.1, 0.9, 1.0}',
+                                vector,
+                                to_tsquery('simple', ?),
+                                8
+                            )
                         )
-                    )
-                    %(order_case_statements)s
-                    DESC,
-                    display_name IS NULL,
-                    avatar_url IS NULL
-                LIMIT ?
-            """ % {
-                "where_clause": where_clause,
-                "order_case_statements": " ".join(additional_ordering_statements),
-            }
-            args = (
-                join_args
-                + (full_query, exact_query, prefix_query)
-                + ordering_arguments
-                + (limit + 1,)
-            )
+                        %(order_case_statements)s
+                        DESC,
+                        display_name IS NULL,
+                        avatar_url IS NULL
+                    LIMIT ?
+                """ % {
+                    "where_clause": where_clause,
+                    "order_case_statements": " ".join(additional_ordering_statements),
+                }
+                args = (
+                    join_args
+                    + (full_query, exact_query, prefix_query)
+                    + ordering_arguments
+                    + (limit + 1,)
+                )
         elif isinstance(self.database_engine, Sqlite3Engine):
             search_query = _parse_query_sqlite(search_term)
 
-            # If enabled, this config option will rank local users higher than those on
-            # remote instances.
-            if self._prefer_local_users_in_search:
-                # This statement checks whether a given user's user ID contains a server name
-                # that matches the local server
-                #
-                # Note that we need to include a comma at the end for valid SQL
-                statement = "user_id LIKE ? DESC,"
-                additional_ordering_statements.append(statement)
+            if not search_term:
+                logger.debug(
+                    "search_user_dir() No search term provided to user_directory, doing plain list"
+                )
+                sql = """
```

Comment:
This query seems to be the same as the one for Postgres.

This function is already a bit messy, but duplicating the same query twice might not be helping things.

I wonder if the structure should be (in rough code):

```python
if not search_term:
    ...
elif isinstance(.., PostgresEngine):
    ...
elif isinstance(.., SqliteEngine):
    ...
```

That might also help to reduce the amount of nesting happening in this function.

Reply:
big +1 to this - it's pretty hard to review as it stands.

I'd actually recommend splitting it into two functions - 200 line functions are no fun.